File: examples/isochromat.cpp
#include <sycomore/isochromat/Model.h>
#include <sycomore/units.h>

#include <xtensor/xview.hpp>

#include <xtensor/xio.hpp>

int main()
{
    using namespace sycomore::units;
    
    sycomore::Vector3R M0{0,0,1};
    // 1D model at a single position. This can take a more complex form, e.g. for
    // a 3D model at two positions: {{0*mm, 0*mm, 0*mm}, {1*mm, 1*mm, 1*mm}} 
    sycomore::TensorQ<2>positions{{0*mm}};
    auto T1 = 1000*ms, T2 = 100*ms, flip_angle = 60*deg, step = 10*ms;
    
    sycomore::isochromat::Model model(T1, T2, M0, positions);
    
    auto pulse = model.build_pulse(flip_angle);
    auto idle = model.build_time_interval(10*ms);
    
    std::vector<std::pair<sycomore::Quantity, sycomore::Vector3R>> record{
        {0*s, xt::view(model.magnetization(), 0)}};
    
    for(std::size_t i=0; i!=10; ++i)
    {
        model.apply(idle);
        record.emplace_back(
            record.back().first+step, xt::view(model.magnetization(), 0));
    }
    
    model.apply(pulse);
    record.emplace_back(
        record.back().first+step, xt::view(model.magnetization(), 0));
    
    for(std::size_t i=0; i!=100; ++i)
    {
        model.apply(idle);
        record.emplace_back(
            record.back().first+step, xt::view(model.magnetization(), 0));
    }
    
    return 0;
}

File: examples/isochromat.py
import sys

import sycomore
from sycomore.units import *

destination, = sys.argv[1:]

M0 = [0,0,1]
positions = [[0*mm, 0*mm, 0*mm]]
T1, T2 = 1000*ms, 100*ms
flip_angle = 60*deg
step = 10*ms

model = sycomore.isochromat.Model(T1, T2, M0, positions)

pulse = model.build_pulse(flip_angle)
idle = model.build_time_interval(10*ms)

record = [[0*s, model.magnetization[0]]]
for _ in range(10):
    model.apply(idle)
    record.append([record[-1][0]+step, model.magnetization[0]])

model.apply(pulse)
record.append([record[-1][0], model.magnetization[0]])

for _ in range(100):
    model.apply(idle)
    record.append([record[-1][0]+step, model.magnetization[0]])

import matplotlib.pyplot
import numpy

time, magnetization = list(zip(*record))
magnetization = numpy.array(magnetization)

x_axis = [x.convert_to(ms) for x in time]
figure, plot = matplotlib.pyplot.subplots(tight_layout=True, figsize=(8, 6))
plot.plot(
    x_axis, numpy.linalg.norm(magnetization[:, :2], axis=-1), label="$M_\perp$")
plot.plot(x_axis, magnetization[:, 2], label="$M_z$")
plot.set(xlim=0, ylim=-0.02, xlabel="Time (ms)", ylabel="$M/M_0$")
matplotlib.pyplot.legend()
matplotlib.pyplot.tight_layout()

figure.savefig(destination)
